test/common/network/connection_impl_test.cc

```diff
@@ -870,6 +887,173 @@ TEST_P(ConnectionImplTest, EmptyReadOnCloseTest) {
   disconnect(true);
 }
 
+// Test that a FlushWrite close immediately triggers a close after the write buffer is flushed.
+TEST_P(ConnectionImplTest, FlushWriteCloseTest) {
+  setUpBasicConnection();
+  connect();
+  // Set a very high timeout value to prevent flaking. We are testing the common case where the
+  // timeout does not trigger.
+  server_connection_->setDelayedCloseTimeout(std::chrono::milliseconds(50000));
+
+  std::shared_ptr<MockReadFilter> client_read_filter(new NiceMock<MockReadFilter>());
+  client_connection_->addReadFilter(client_read_filter);
+
+  NiceMockConnectionStats stats;
+  server_connection_->setConnectionStats(stats.toBufferStats());
+
+  Buffer::OwnedImpl data("data");
+  server_connection_->write(data, false);
+
+  // Server connection flushes the write and immediately closes the socket.
+  // There shouldn't be a read/close race here (see issue #2929), since the client is blocked on
+  // reading and the connection should close gracefully via FIN.
+  EXPECT_CALL(*client_read_filter, onData(BufferStringEqual("data"), false))
+      .Times(1)
+      .WillOnce(InvokeWithoutArgs([&]() -> FilterStatus {
+        dispatcher_->exit();
+        return FilterStatus::StopIteration;
+      }));
+  EXPECT_CALL(stats.delayed_close_timeouts_, inc()).Times(0);
+  EXPECT_CALL(client_callbacks_, onEvent(ConnectionEvent::RemoteClose)).Times(1);
+  EXPECT_CALL(server_callbacks_, onEvent(ConnectionEvent::LocalClose)).Times(1);
+  server_connection_->close(ConnectionCloseType::FlushWrite);
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
+}
+
+// Test that a FlushWrite close will trigger a timeout which closes the connection when the write
+// buffer is not flushed within the configured interval.
+TEST_P(ConnectionImplTest, FlushWriteCloseTimeoutTest) {
+  NiceMock<Event::MockDispatcher> dispatcher;
+  EXPECT_CALL(dispatcher.buffer_factory_, create_(_, _))
+      .WillRepeatedly(Invoke([](std::function<void()> below_low,
+                                std::function<void()> above_high) -> Buffer::Instance* {
+        return new Buffer::WatermarkBuffer(below_low, above_high);
+      }));
+
+  auto timer = new Event::MockTimer(&dispatcher);
```

Comment:
Why can't we put these on the stack? They shouldn't be needed after the TEST_P body ends, right?

Reply:
The WatermarkBuffer and MockTimer need to be allocated on the free store because ownership is handed over to Envoy::Network::ConnectionImpl when Envoy::MockBufferFactory::create() and Envoy::Event::MockDispatcher::createTimer() are called (respectively).

Reply:
That explanation makes sense. I get nervous when I see bare new calls, especially without a comment. I think this test might be doing the opposite of the pattern I usually see, which is to have a `unique_ptr` here in the test body which is `std::move`d into the right place to be returned by `Envoy::MockBufferFactory::create()` but only after you steal a bare pointer out of the `unique_ptr` to use when setting up any the gmock expectations. This is a clear case of dangerous cheating the `unique_ptr`, but it's ok in this limited test case example since we know the unique pointer scope lives as long as the body of this test. I think it leaves the object life cycle a bit clearer defined than if you don't use `unique_ptr` here.

If this doesn't work in your case, please add a comment explaining the object lifecycle at the bare `new`s. Happy to chat more about this in person tomorrow too if it helps.